src/config/Config.ts

```diff
@@ -15,34 +15,113 @@
+import {SerializationConfig, SerializationConfigImpl} from './SerializationConfig';
 import {Statistics} from '../statistics/Statistics';
 import {LogLevel} from '..';
 import {ILogger} from '../logging/ILogger';
 import {JsonStringDeserializationPolicy} from './JsonStringDeserializationPolicy';
-import {ConnectionStrategyConfig, ReconnectMode} from './ConnectionStrategyConfig';
-import {LoadBalancer} from '../LoadBalancer';
+import {ConnectionStrategyConfig, ConnectionStrategyConfigImpl, ReconnectMode} from './ConnectionStrategyConfig';
+import {LoadBalancerType, LoadBalancerConfig, LoadBalancerConfigImpl} from './LoadBalancerConfig';
 import {IndexConfig} from './IndexConfig';
 import {IndexType} from './IndexType';
 import {ConnectionRetryConfig} from './ConnectionRetryConfig';
 
-const DEFAULT_CLUSTER_NAME = 'dev';
-
 /**
- * Top level configuration object of Hazelcast client. Other configurations items are properties of this object.
+ * Top level configuration object of Hazelcast client.
+ * Other configurations items are properties of this object.
  */
-export class ClientConfig {
+export interface ClientConfig {
+
+    /**
+     * Name of the cluster to connect to. By default, set to `dev`.
+     */
+    clusterName?: string;
+
+    /**
+     * Name of the client instance. By default, set to `hz.client_${CLIENT_ID}`, where
+     * `CLIENT_ID` starts from `0` and it is incremented by `1` for each new client.
+     */
+    instanceName?: string;
+
+    /**
+     * Labels for the client to be sent to the cluster.
+     */
+    clientLabels?: string[];
+
+    /**
+     * Network config of the client.
+     */
+    network?: ClientNetworkConfig;
+
+    /**
+     * Connection strategy config of the client.
+     */
+    connectionStrategy?: ConnectionStrategyConfig;
+
+    /**
+     * Load balancer config for the client.
+     */
+    loadBalancer?: LoadBalancerConfig;
+
+    /**
+     * Lifecycle listeners to be attached to the client.
+     */
+    lifecycleListeners?: Array<(state: LifecycleState) => void>;
+
+    /**
+     * Membership listeners to be attached to the client.
+     */
+    membershipListeners?: MembershipListener[];
+
+    /**
+     * User-defined serialization config for the client.
+     */
+    serialization?: SerializationConfig;
+
+    /**
+     * Near Cache config for the client.
+     */
+    nearCaches?: { [name: string]: NearCacheConfig };
+
+    /**
+     * Configs for ReliableTopics.
+     */
+    reliableTopics?: { [name: string]: ReliableTopicConfig };
+
+    /**
+     * Configs for FlakeIdGenerators.
+     */
+    flakeIdGenerators?: { [name: string]: FlakeIdGeneratorConfig };
+
+    /**
+     * Custom logger implementation for the client.
+     */
+    customLogger?: ILogger;
+
+    /**
+     * Custom credentials to be used as a part of authentication on the cluster.
+     */
+    customCredentials?: any;
+
+    /**
+     * User-defined properties.
+     */
+    properties?: Properties;
```

Comment:
What do you think about making available options in the `properties` top-level configuration elements? Like `heartbeatInterval`, `heartbeatTimeout` etc. I believe they are hard to use now (requires lots of typing and hard to remember) but moving them to top-level may pollute the config options.

Reply:
I'd rather leave them as is, as it's a rare case when a user needs to change one of those. Of course, they should be documented well.